tools/ci_connector_ops/ci_connector_ops/pipelines/publish.py

```diff
@@ -0,0 +1,232 @@
+
+        specs_to_uploads: List[Tuple[str, File]] = [(self.oss_spec_key, self._get_spec_as_file(oss_spec))]
+
+        if oss_spec != cloud_spec:
+            specs_to_uploads.append(self.cloud_spec_key, self._get_spec_as_file(cloud_spec, "cloud_spec_to_cache.json"))
+
+        upload_results = []
+        for key, file in specs_to_uploads:
+            exit_code, stdout, stderr = await upload_to_gcs(
+                self.context.dagger_client, file, key, self.spec_bucket_name, self.gcs_credentials
+            )
+            upload_results.append(StepResult(self, status=StepStatus.from_exit_code(exit_code), stdout=stdout, stderr=stderr))
+        return upload_results
+
+
+async def run_connector_publish_pipeline(
+    context: ConnectorContext, semaphore: anyio.Semaphore, pre_release: bool, spec_bucket_name: str, metadata_bucket_name: str
+) -> ConnectorReport:
+    """Run a publish pipeline for a single connector.
+
+    1. Validate the metadata file.
+    2. Check if the connector image already exists.
+    3. Build the connector, with platform variants.
+    4. Upload its spec to the spec cache bucket.
+    5. Push the connector to DockerHub, with platform variants.
+    6. Upload its metadata file to the metadata service bucket.
+
+    Returns:
+        ConnectorReport: The reports holding publish results.
+    """
+    async with semaphore:
+        async with context:
+            report_name = "PUBLISH RESULTS"
+            spec_cache_service_account: dagger.Secret = context.dagger_client.set_secret(
+                "spec_cache_service_account_key", os.environ["SPEC_CACHE_SERVICE_ACCOUNT_KEY"]
+            )
+            metadata_service_account: dagger.Secret = context.dagger_client.set_secret(
+                "metadata_service_account_key", os.environ["METADATA_SERVICE_ACCOUNT_KEY"]
+            )
+
+            metadata_validation_step = metadata.MetadataValidation(context, context.metadata_path)
+            check_connector_image_does_not_exists_step = CheckConnectorImageDoesNotExist(context, pre_release)
+            build_connector_step = BuildConnectorForPublish(context)
+            upload_spec_to_cache_step = UploadSpecToCache(context, pre_release, spec_bucket_name, spec_cache_service_account)
+            push_connector_image_to_registry_step = PushConnectorImageToRegistry(context, pre_release)
+            metadata_upload_step = metadata.MetadataUpload(
+                context, context.metadata_path, metadata_bucket_name, await metadata_service_account.plaintext()
+            )
+            step_results = []
+
+            # TODO this sequence of "return on Failure" is not beautiful.
+            # We'll improve this soon with a different and global approach to chain steps.
+            metadata_validation_result = await metadata_validation_step.run()
+            step_results.append(metadata_validation_result)
+            if metadata_validation_result.status is StepStatus.FAILURE:
```

Comment:
Im thinking that all these return on failures could benefit from the skip on failure logic in the [run_steps](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/pipelines/metadata.py#L22) function in the metadata pipeline.

What do you think of extending that function to work here. Would be a good step in the direction of the global approach you mention

Reply:
Yes you're right. I opened a side PR to adapt it for connector use cases and use it https://github.com/airbytehq/airbyte/pull/25422